docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
-     0  0  3
-     4  0  5
-    [torch.FloatTensor of size 2x3]
-
-You can also construct hybrid sparse tensors, where only the first n
-dimensions are sparse, and the rest of the dimensions are dense.
-
-    >>> i = torch.LongTensor([[2, 4]])
-    >>> v = torch.FloatTensor([[1, 3], [5, 7]])
-    >>> torch.sparse.FloatTensor(i, v).to_dense()
-     0  0
-     0  0
-     1  3
-     0  0
-     5  7
-    [torch.FloatTensor of size 5x2]
+    >>> i = [[0, 2], [1, 0], [1, 2]]
+    >>> v =  [3,      4,      5    ]
+    >>> s = torch.sparse_coo_tensor(list(zip(*i)), v, (2, 3))
+    >>> # Or another equivalent formulation to get s
+    >>> s = torch.sparse_coo_tensor(torch.tensor(i).t(), v, (2, 3))
+    >>> torch.sparse_coo_tensor(i.t(), v, torch.Size([2,3])).to_dense()
+    tensor([[0, 0, 3],
+            [4, 0, 5]])
+
+Sparse COO tensors support indexing operations (except slicing in
+sparse dimensions):
+
+    >>> s[1]
+    tensor(indices=tensor([[0, 2]]),
+           values=tensor(  [4, 5]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+    >>> s[1, 0]
+    tensor(4)
 
 An empty sparse tensor can be constructed by specifying its size:
 
-    >>> torch.sparse.FloatTensor(2, 3)
-    SparseFloatTensor of size 2x3 with indices:
-    [torch.LongTensor with no dimension]
-    and values:
-    [torch.FloatTensor with no dimension]
+    >>> torch.sparse_coo_tensor(size=(2, 3))
+    tensor(indices=tensor([], size=(2, 0)),
+           values=tensor([], size=(0,)),
+           size=(2, 3), nnz=0, layout=torch.sparse_coo)
+
+.. _sparse-hybrid-coo-docs:
+
+Hybrid COO tensors
+------------------
+
+You can also construct hybrid tensors, where only the first dimensions
+are sparse, and the rest of the dimensions are dense. For instance,
+
+    >>> i = [[2, 4]]
+    >>> v = [[1, 3], [5, 7]]
+    >>> s = torch.sparse_coo_tensor(i, v, (5, 2))
+    >>> s.to_dense()
+    tensor([[0, 0],
+            [0, 0],
+            [1, 3],
+            [0, 0],
+            [5, 7]])
+
+where ``s`` is a (1+1)-dimensional :ref:`hybrid tensor
+<sparse-hybrid-docs>`:
+
+    >>> s[0]
+    tensor([0, 0])
+    >>> s[2]
+    tensor([1, 3])
+
+Notice that slicing in dense dimensions is supported (except with
+negative step):
+
+    >>> s[2, 1:]
+    tensor([3])
+
+In general, a (M + K)-dimensional coalesced sparse COO tensor ``s``
+with the number of specified elements ``nse`` has the following
+invariants:
+
+    >>> s = s.coalesce()
+    >>> M, K = s.sparse_dim(), s.dense_dim()
+    >>> assert M + K == len(s.shape) == s.ndim
+    >>> assert s.indices().shape == (M, nse)
+    >>> assert s.values().shape == (nse,) + s.shape[M:]
+
+Since ``s.indices()`` is always a 2D tensor, the smallest sparse
+dimension M = 1.  Therefore, representation of a sparse tensor of M =
+0 is simply a dense tensor.
+
+.. _sparse-uncoalesed-coo-docs:
+
+Uncoalesced sparse COO tensors
+------------------------------
+
+Torch sparse COO tensor format permits *uncoalesced* sparse tensors,
```

Comment:
Per the above comments, I wonder if this should be included in the mathematical description of the coordinate list format?

Reply:
See the new text. The concept of "uncoalesced tensor" ought to be an implementation detail.